Implement Physical Server Toolbar actions with API

AparnaKarve opened the pull request:
The Physical Server Toolbar actions like `power_on`, `power_off` etc are fired directly from the Toolbar itself via REST API called from the angular controller.

Works in both `show_list` and `show` screens

-------------
`show_list` -- Button action applied to multiple selected servers

<img width="1277" alt="screen shot 2017-05-30 at 2 41 29 pm" src="https://cloud.githubusercontent.com/assets/1538216/26606383/9155b72e-4546-11e7-8948-fa219573da51.png">


---------------
`show` -- Button action applied to the selected server

<img width="1275" alt="screen shot 2017-05-30 at 2 44 30 pm" src="https://cloud.githubusercontent.com/assets/1538216/26606407/a10d5dca-4546-11e7-8750-2ad3978558ea.png">

AparnaKarve commented:
@GabrielSVinha Could you please test this in your environment? Thanks.

AparnaKarve commented:
@miq-bot add_label fine/no,enhancement,compute/physical infrastructure

GabrielSVinha commented:
@AparnaKarve I will do it. So, the actions on physical server controller are dismissed?

AparnaKarve commented:
Yes, that is correct. We do not need any Rails code for the `button` action anymore.

himdel commented:
So, uh.. a couple of questions...

   * why is this even an angular controller? It doesn't have views, does not render anything... looks like what you're doing is just listen to an observable and do API requests, and you don't need angular for that.

   * what's the deal with `vm.physicalServerToolbarModel.server`? Looks like you're listening every time an item is checked and manually keeping a list of checked items .. but the toolbar already supports reading `ManageIQ.gridChecks`, so you don't have to keep track manually.

   * supported actions - is it really necessary? Because, this just adds another place one has to add the button for it to work, as opposed to every other controller..

   * the combination of `miqFlashLater` + `miqFlashSaved` right after each other doesn't make sense, that's just `add_flash`. (Or you can `miqFlashSaved` only once, outside the loop, it not re-rendering N times was the intent.)

   * sending only `type` in `function-data` means any controller will get those messages .. add something specific to this controller please - most places seem to have settled on using `controller: 'fooBarController' in data, so that each (js) controller can just check if the message is for meant for it by checking that the name matches

  * seems like you no longer need most of the other entries on those toolbar buttons, delete please :) (For example, `url_parms` is only used here so that the toolbar code sends the list of checked items, which you no longer need.)

  * `#form` could be a little less generic too :)

AparnaKarve commented:
> why is this even an angular controller? It doesn't have views, does not render anything... looks like what you're doing is just listen to an observable and do API requests, and you don't need angular for that.

It currently does not have any views, but there will be views down the line.
So what would be the right course of action here?
 - Have a regular JS source that listens to an observable and does API requests for all actions that don't render a view 
- And have an angular controller only when there is a view?
(One notable advantage of having an angular controller here at this point, is that, I can directly inject the Physical Server id, which I think is the cleanest approach rather than trying to decipher the id from the url or having it as a hidden element in the form. What's your take on that?)

> what's the deal with vm.physicalServerToolbarModel.server? Looks like you're listening every time an item is checked and manually keeping a list of checked items .. but the toolbar already supports reading ManageIQ.gridChecks, so you don't have to keep track manually.

Yes, that was exactly the deal to keep track of checked items.
Thanks for the tip on `ManageIQ.gridChecks`

> the combination of miqFlashLater + miqFlashSaved right after each other doesn't make sense, that's just add_flash. (Or you can miqFlashSaved only once, outside the loop, it not re-rendering N times was the intent.)

My original intent *was* to render the flash N times, at least for the first pass, while we are still in the testing phase. Will change it appropriately at the right time.

AparnaKarve commented:
@himdel The other advantage of using an angular controller (even though the controller is not dealing with any views right now), is consistency... consistency with how we make the API calls.
```javascript
API.post(url)
  .then(postFunction)
  .catch(miqService.handleFailure)
```

I *really* like that consistency.

AparnaKarve commented:
> sending only type in function-data means any controller will get those messages .. add something specific to this controller please - most places seem to have settled on using `controller: 'fooBarController' in data, so that each (js) controller can just check if the message is for meant for it by checking that the name matches

Somehow I don't see this happening.

All messages triggered by the Physical Server toolbar are only redirected to the observable in the angular controller `physicalServerToolbarController`.

himdel commented:
I do agree that the controller makes some things much simpler, you *could* get at the value by reading `ManageIQ.record.recordId`, and you *could* use `vanillaJsAPI` instead of `API`, but I guess that's problematic too..

Like, it doesn't make *much* sense to me to ignore angular when we can use it, so I guess you convinced me :).

But can you make it into a component please? Just so that we don't have to convert it again..

> All messages triggered by the Physical Server toolbar are only directed to the observable in the angular controller physicalServerToolbarController.

Sorry, not true, there is only one observable, used for all the messages. And given that right now, any message with a `type` fileld triggers an API request.... Not that nice :) (Though to be fair, the only thing that generates events with `type` right now is the middleware toolbar, so it should not happen right now.)

But.. you know, way easier to add some extra field now, than to have to unravel it later ;). (I'm not really married to the idea of adding `controller: 'physicalServerToolbarController'`, it makes sense because searchable, but if you decide to add something simpler, anything that somehow indicates this is about physical servers is fine :)).

AparnaKarve commented:
> you know, way easier to add some extra field now, than to have to unravel it later

Adding an extra field like `controller: 'physicalServerToolbarController'` is no problem at all and I intend to do it.

I suppose `ManageIQ.angular.rxSubject.subscribe` acts like a "global observable", so to speak, only in a regular vanilla JS code. But within the confines of an angular controller, that's not the case -- the angular controller acts like a natural filter for the observable.
It would be nice if we could convert all vanilla JS code that uses `ManageIQ.angular.rxSubject.subscribe` to an angular controller.

> can you make it into a component please? 

Sure, I was thinking about that too.

AparnaKarve commented:
/cc @martinpovolny

GabrielSVinha commented:
Tested, that works fine in physical appliance.

martinpovolny commented on app/assets/javascripts/components/physical_infrastructures/physical-server-toolbar.js:
```diff
@@ -0,0 +1,50 @@
+  bindings: {
+    physicalServerId: '=?',
+  },
+  controllerAs: 'vm',
```
what does "vm" stand for here?

martinpovolny replied:
should it be (here and below) rather "toolbar"?

AparnaKarve replied:
`vm` is the standard abbreviation for `view model` (I believe).
In the angular community, it is almost always used in the context of `controllerAs`.

I suppose we could call it whatever we want.
Though, I just don't know if we should.

martinpovolny replied:
Oh, in that case we surely should. As we don't call model instances "model" in Rails code or we don't call bool variables "bool" or "boolean" or in pascal we don't call a string variable a "string" unless we are writing a tutorial on using different types and want to point out the difference ;-)

AparnaKarve replied:
True. Although it just seems a bit odd to call `vm` with a different name. Probably because I have never come across any code that uses a different naming convention.

Anyway, per your request, I will be changing it to `toolbar`
We can always change it back to `vm` if requried ;-)

AparnaKarve replied:
Addressed this here: https://github.com/ManageIQ/manageiq-ui-classic/pull/1380/commits/c6c937d2f91f569fb65eee229308aa25f7412c41

martinpovolny commented on app/assets/javascripts/components/physical_infrastructures/physical-server-toolbar.js:
```diff
@@ -0,0 +1,50 @@
+      vm.action = event.type;
+
+      if (vm.action) {
+        vm.servers = [];
```
this line is needed?

AparnaKarve replied:
That line can be eliminated.

AparnaKarve replied:
Done. https://github.com/ManageIQ/manageiq-ui-classic/pull/1380/commits/c6c937d2f91f569fb65eee229308aa25f7412c41

martinpovolny commented on app/assets/javascripts/services/miq_service.js:
```diff
@@ -53,6 +54,14 @@ ManageIQ.angular.app.service('miqService', ['$timeout', '$document', '$q', funct
     $('#flash_msg_div').text("");
   };
 
+  this.miqFlashLater = function(msgObj) {
```
why do we need this method? all it does is call the global one, can we avoid adding this?

AparnaKarve replied:
A couple of things are happening here.

`miqService` is our generic helper service that mostly contains common code that the angular controllers consume. Flash messages happens to be one part of the common code. 

To be able to use `miqFlashLater` in the physical server toolbar component, I had to declare it as a function in `miqService`.
By doing this, 
- I have added one more Flash function to the existing family of Flash functions in `miqService`
- I have added `miqFlashLater` in the global comment here https://github.com/AparnaKarve/manageiq-ui-classic/blob/0d7737978e05fc9e1a45f8b56371c6889151fe7b/app/assets/javascripts/services/miq_service.js#L2, which actually addresses a CC error
Note that, CC discourages the use of global methods in angular code and this is the only way to keep the CC error at bay.

martinpovolny replied:
I understand, right so this makes sense. I had the impression that declaring a variable as global as it e.g. in `app/assets/javascripts/services/post_service.js` : 
```
/* global miqFlashLater */
```
is the way to deal with this problem (make CC and similar tools happy)

Yes, but from clean code perspective, that you did is surely better. I'd just love to have it done the same way everywhere (so probably `postservice.js`) needs fixing.

martinpovolny commented on app/views/physical_server/show.html.haml:
```diff
@@ -1,5 +1,12 @@
+- @angular_form = true
```
I see this on places and I wonder.

This is to skip the content of `app/views/layouts/_x_edit_buttons.html.haml` or does it have some more meaning?

I am wondering on it's purpose since setting variables inside a view. 

We are actually counting on that it leaks somewhere. Help me understant why is this needed maybe I can help to figure out some better way to do it.

AparnaKarve replied:
Skipping `app/views/layouts/_x_edit_buttons.html.haml` seems to be the only real purpose.

And since this is not an Add/Edit form, the question of rendering buttons does not arise here.
This declaration `- @angular_form = true` can be removed from here.

martinpovolny replied:
Cool so please, remove it we should not be adding lines that are not needed. It makes future refactoring work more difficult.

Thanks for your explanations.

I am happy that at least some of my comment make sense with my limited Angular skills ;-)

AparnaKarve replied:
Done. https://github.com/ManageIQ/manageiq-ui-classic/pull/1380/commits/c6c937d2f91f569fb65eee229308aa25f7412c41

AparnaKarve commented:
@martinpovolny Any idea what could be done with the CC errors `Similar code found in 1 other location (mass = 84)`? They seem to be appear every time we make a change in the toolbar code.
Those are pretty much the only errors here.

AparnaKarve commented:
I believe I have addressed all concerns/questions here and this seems like a really efficient solution to call REST API code directly from the Toolbar via an angular component.
Using Angular as opposed to Vanialla JS has many advantages as @himdel and I have discussed above.

@GabrielSVinha Thanks for testing this.

@dclarizio Please review. Thanks.

martinpovolny commented:
> @martinpovolny Any idea what could be done with the CC errors Similar code found in 1 other location (mass = 84)? They seem to be appearing every time we make a change in the toolbar code.
Those are pretty much the only errors here.

Probably nothing at this point. I'd like to turn that off for toolbars.
